crates/c-api/src/func.rs

```diff
@@ -217,8 +219,13 @@ fn _wasmtime_func_call(
     let result = panic::catch_unwind(AssertUnwindSafe(|| func.call(&params)));
     match result {
         Ok(Ok(out)) => {
-            for (slot, val) in results.iter_mut().zip(out.iter()) {
-                *slot = wasm_val_t::from_val(val);
+            for (slot, val) in results.iter_mut().zip(out.into_vec().into_iter()) {
+                unsafe {
+                    // NB: The results array is likely uninitialized memory, so
+                    // use `ptr::write` rather than assignment (which tries to
+                    // run destructors).
+                    ptr::write(slot.as_mut_ptr(), wasm_val_t::from_val(val));
```

Comment:
Could you add a fixme to use the `write` method of `MaybeUninit` once it is stabilized?